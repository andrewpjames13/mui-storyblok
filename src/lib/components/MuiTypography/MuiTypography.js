import React, { createElement } from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';

import Storyblok from '../../utils/Storyblok';
import MuiIcon from '../MuiIcon/MuiIcon';
import MuiIconButtonRedirect from '../MuiIconButtonRedirect/MuiIconButtonRedirect';
import MuiIconButtonHref from '../MuiIconButtonHref/MuiIconButtonHref';
import MuiIconButtonDownload from '../MuiIconButtonDownload/MuiIconButtonDownload';
import MuiIconButtonDialog from '../MuiIconButtonDialog/MuiIconButtonDialog';
import MuiButtonRedirect from '../MuiButtonRedirect/MuiButtonRedirect';
import MuiButtonHref from '../MuiButtonHref/MuiButtonHref';
import MuiButtonDownload from '../MuiButtonDownload/MuiButtonDownload';
import MuiButtonDialog from '../MuiButtonDialog/MuiButtonDialog';
import MuiTooltip from '../MuiTooltip/MuiTooltip';
import MuiLink from './components/MuiLink/MuiLink';

/**
 * MuiTypography is used in storyblok redirect to react routes
 */

export const MuiTypography = ({
  align,
  color,
  variant,
  height,
  width,
  rootClass,
  content,
}) => {
  const components = {
    MuiLink,
    MuiIcon,
    MuiIconButtonRedirect,
    MuiIconButtonHref,
    MuiIconButtonDownload,
    MuiIconButtonDialog,
    MuiButtonRedirect,
    MuiButtonHref,
    MuiButtonDownload,
    MuiButtonDialog,
    MuiTooltip,
  };

  const defaultStyles = {
    height,
    width,
  };
  const styles = Storyblok.arrayToMuiStyles(rootClass, defaultStyles);

  return (
    <Typography
      className={styles.root}
      align={align}
      color={color}
      component="div"
    >
      {content.map((item, index) => {
        if (item.component === 'MuiText') {
          const style = Storyblok.arrayToMuiStyles(item.rootClass);
          return (
            <Typography
              component="span"
              key={index}
              className={style.root}
              variant={item.variant}
              align={item.align}
              color={item.color}
            >
              {` ${item.text} `}
            </Typography>
          );
        }
        return createElement(components[item.component], Object.assign(item, { key: index }));
      })}
    </Typography>
  );
};

export default MuiTypography;

MuiTypography.propTypes = {
  /**
   * stroyblok multiselect of css classes
   * Override or extend the styles applied to the component
   * */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /**
   * mui prop: 'inherit' | 'left' | 'center' | 'right' | 'justify'
   * Set the text-align on the component.
   * */
  align: PropTypes.string,
  /**
   * mui prop: 'initial'| 'inherit'| 'primary'| 'secondary'| 'textPrimary'| 'textSecondary'| 'error'
   * The color of the component. It supports those theme colors that make sense for this component.
   * */
  color: PropTypes.string,
  /**
   * mui prop: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'subtitle1' | 'subtitle2' | 'body1' | 'body2' | 'caption' | 'button' | 'overline' | 'srOnly' | 'inherit'
   * Applies the theme typography styles.
   * */
  variant: PropTypes.string,
  /** hight of the Typography element */
  height: PropTypes.string,
  /** width of the Typography element */
  width: PropTypes.string,

  content: PropTypes.arrayOf(PropTypes.shape({
    component: PropTypes.string.isRequired,
  })).isRequired,
};

MuiTypography.defaultProps = {
  width: '100%',
  height: '100%',
  align: 'inherit',
  color: 'initial',
  variant: 'body1',
  rootClass: [],
};
