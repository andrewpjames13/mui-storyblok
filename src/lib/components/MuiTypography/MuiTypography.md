MuiTypography nested icon/text example:

```js
const muiText = {
    component: 'MuiTypography',
    content: [
      {
      component: 'MuiText',
      content: 'you can concatenate strings',
      },
      {
      component: 'MuiText',
      content: ' like this',
      color: 'secondary',
      },
    ],
    height: '100%',
    variant: 'h3',
    width: '100%',
    color: 'inherit',
}

const icon = {
    component: 'MuiIcon',
    iconName: 'done_all',
};

const iconButtonDownload = {
    component: 'MuiIconButtonDownload',
    href: "https://miro.medium.com/max/1200/1*mk1-6aYaf_Bes1E3Imhc0A.jpeg",
    fileName: "cool.jpeg",
    iconButton: [{
      component: 'MuiIconButton',
      icon: [icon],
    }],
};

<MuiTypography {...muiText} />
```

MuiTypography Variants example:

```js
const h1 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'H1. Heading',
    variant: 'h1',
  }],
};

const h2 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'H2. Heading',
    variant: 'h2',
  }],
};

const h3 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'H3. Heading',
    variant: 'h3',
  }],
};

const h4 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'H4. Heading',
    variant: 'h4',
  }],
};

const h5 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'H5. Heading',
    variant: 'h5',
  }],
};

const h6 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'H6. Heading',
    variant: 'h6',
  }],
};

const subtitle1 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'subtitle 1. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Quos blanditiis tenetur',
    variant: 'subtitle1',
  }],
};

const subtitle2 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'subtitle 2. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Quos blanditiis tenetur',
    variant: 'subtitle2',
  }],
};

const body1 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'body 1. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Quos blanditiis tenetur unde suscipit, quam beatae rerum inventore consectetur, neque doloribus, cupiditate numquam dignissimos laborum fugiat deleniti? Eum quasi quidem quibusdam.',
    variant: 'body1',
  }],
};

const body2 = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'body 2. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Quos blanditiis tenetur unde suscipit, quam beatae rerum inventore consectetur, neque doloribus, cupiditate numquam dignissimos laborum fugiat deleniti? Eum quasi quidem quibusdam.',
    variant: 'body2',
  }],
};

const buttonVariant = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'button text',
    variant: 'button',
  }],
};

const caption = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'caption text',
    variant: 'caption',
  }],
};

const overline = {
  component: 'MuiTypography',
  content: [{
    component: 'MuiText',
    content: 'overline text',
    variant: 'overline',
  }],
};


<div>
  <MuiTypography {...h1} />
  <MuiTypography {...h2} />
  <MuiTypography {...h3} />
  <MuiTypography {...h4} />
  <MuiTypography {...h5} />
  <MuiTypography {...h6} />
  <MuiTypography {...subtitle1} />
  <MuiTypography {...subtitle2} />
  <MuiTypography {...body1} />
  <MuiTypography {...body2} />
  <MuiTypography {...buttonVariant} />
  <MuiTypography {...caption} />
  <MuiTypography {...overline} />
</div>
```