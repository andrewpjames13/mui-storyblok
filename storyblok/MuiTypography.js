module.exports = {
  "name": "MuiTypography",
  "display_name": null,
  "updated_at": "2020-05-29T17:04:32.640Z",
  "schema": {
    "align": {
      "type": "option",
      "use_uuid": true,
      "options": [
        {
          "value": "inherit",
          "name": "inherit"
        },
        {
          "value": "left",
          "name": "left"
        },
        {
          "value": "center",
          "name": "center"
        },
        {
          "value": "right",
          "name": "right"
        },
        {
          "value": "justify",
          "name": "justify"
        }
      ],
      "default_value": "inherit",
      "required": true
    },
    "color": {
      "type": "option",
      "use_uuid": true,
      "options": [
        {
          "value": "initial",
          "name": "initial"
        },
        {
          "value": "inherit",
          "name": "inherit"
        },
        {
          "value": "primary",
          "name": "primary"
        },
        {
          "value": "secondary",
          "name": "secondary"
        },
        {
          "value": "textPrimary",
          "name": "textPrimary"
        },
        {
          "value": "textSecondary",
          "name": "textSecondary"
        },
        {
          "value": "error",
          "name": "error"
        }
      ],
      "default_value": "initial",
      "required": true
    },
    "variant": {
      "type": "option",
      "use_uuid": true,
      "options": [
        {
          "value": "h1",
          "name": "h1"
        },
        {
          "value": "h2",
          "name": "h2"
        },
        {
          "value": "h3",
          "name": "h3"
        },
        {
          "value": "h4",
          "name": "h4"
        },
        {
          "value": "h5",
          "name": "h5"
        },
        {
          "value": "h6",
          "name": "h6"
        },
        {
          "value": "subtitle1",
          "name": "subtitle1"
        },
        {
          "value": "subtitle2",
          "name": "subtitle2"
        },
        {
          "value": "body1",
          "name": "body1"
        },
        {
          "value": "body2",
          "name": "body2"
        },
        {
          "value": "caption",
          "name": "caption"
        },
        {
          "value": "button",
          "name": "button"
        },
        {
          "value": "overline",
          "name": "overline"
        },
        {
          "value": "srOnly",
          "name": "srOnly"
        },
        {
          "value": "inherit",
          "name": "inherit"
        }
      ],
      "default_value": "body1",
      "required": true
    },
    "rootClass": {
      "type": "options",
      "options": [
        {
          "value": "color: white",
          "name": "color white"
        }
      ],
      "folder_slug": "",
      "description": "",
      "display_name": "",
      "translatable": false,
      "required": false,
      "external_datasource": ""
    },
    "width": {
      "type": "text",
      "required": true,
      "default_value": "100%"
    },
    "height": {
      "type": "text",
      "required": true,
      "default_value": "100%"
    },
    "content": {
      "type": "bloks",
      "maximum": "",
      "restrict_components": true,
      "component_whitelist": [
        "MuiLink",
        "MuiIcon",
        "MuiIconButtonRedirect",
        "MuiIconButtonHref",
        "MuiIconButtonDownload",
        "MuiIconButtonDialog",
        "MuiButtonRedirect",
        "MuiButtonDownload",
        "MuiButtonDialog",
        "MuiTooltip",
        "MuiButtonHref",
        "MuiText"
      ],
      "required": true
    }
  },
  "image": null,
  "preview_field": null,
  "is_root": false,
  "preview_tmpl": null,
  "is_nestable": true,
  "all_presets": [],
  "preset_id": null,
  "real_name": "MuiTypography",
  "component_group_uuid": null
}