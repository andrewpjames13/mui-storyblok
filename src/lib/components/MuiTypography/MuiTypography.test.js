import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import MuiTypography from './MuiTypography';

function setup() {
  const props = {
    content: [{
      component: 'MuiText',
      text: 'this is text',
    }, {
      component: 'MuiLink',
      text: 'btn',
      href: 'google.com',
    }, {
      component: 'MuiIconButtonRedirect',
      redirectRoute: 'page-welcom',
      iconButton: [{
        component: 'MuiIconButton',
        icon: [{
          component: 'MuiIcon',
          iconName: 'star',
        }],
      }],
    }, {
      component: 'MuiIcon',
      iconName: 'android',
    }],
  };
  const comp = shallow(<MuiTypography {...props} />);
  return { comp, props };
}

describe('<MuiTypography />', () => {
  it('renders MuiTypography', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();

    const tree = renderer.create(<MuiTypography {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
