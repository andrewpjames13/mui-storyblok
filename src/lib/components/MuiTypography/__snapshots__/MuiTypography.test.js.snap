// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiTypography /> snapshot 1`] = `
<p
  className="MuiTypography-root makeStyles-root-3 MuiTypography-body1"
>
  this is text
  <a
    className="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-root-34 MuiTypography-colorPrimary"
    href="#"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
  >
    btn
  </a>
  <button
    className="MuiButtonBase-root MuiIconButton-root makeStyles-root-41"
    data-testid="muiIconButton"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiIconButton-label"
    >
      <span
        aria-hidden={true}
        className="material-icons MuiIcon-root makeStyles-root-54 MuiIcon-colorSecondary"
      >
        star
      </span>
    </span>
  </button>
  <span
    aria-hidden={true}
    className="material-icons MuiIcon-root makeStyles-root-64 MuiIcon-colorSecondary"
  >
    android
  </span>
</p>
`;
